src/expand.rs

```diff
@@ -95,9 +102,236 @@ fn impl_struct(input: &DeriveInput, data: &DataStruct) -> Result<TokenStream> {
+        });
+        segments.push_value(PathSegment {
+            ident: Ident::new(name, Span::call_site()),
+            arguments: PathArguments::None,
+        });
+    }
+    Path {
+        leading_colon: if leading_colon {
+            Some(Colon2 {
+                spans: [Span::call_site(), Span::mixed_site()],
+            })
+        } else {
+            None
+        },
+        segments,
+    }
+}
+
+/// Add a requirement for [core::fmt::Display] to a `where` predicate for some type.
+fn add_display_constraint_to_type_predicate(
+    predicate_that_needs_a_display_impl: &mut PredicateType,
+) {
+    let display_path = join_paths(&["core", "fmt", "Display"], false);
+
+    let display_bound = TypeParamBound::Trait(TraitBound {
+        paren_token: None,
+        modifier: TraitBoundModifier::None,
+        lifetimes: None,
+        path: display_path,
+    });
+    if !predicate_that_needs_a_display_impl.bounds.is_empty() {
+        predicate_that_needs_a_display_impl.bounds.push_punct(Add {
+            spans: [Span::call_site()],
+        });
+    }
+
+    predicate_that_needs_a_display_impl
+        .bounds
+        .push_value(display_bound);
+}
+
+/// Map each declared generic type parameter to the set of all trait boundaries declared on it.
+///
+/// These boundaries may come from the declaration site:
+///     pub enum E<T: MyTrait> { ... }
+/// or a `where` clause after the parameter declarations:
+///     pub enum E<T> where T: MyTrait { ... }
+/// This method will return the boundaries from both of those cases.
+fn extract_trait_constraints_from_source(
+    where_clause: &WhereClause,
+    type_params: &[&TypeParam],
+) -> Vec<(Ident, Vec<TraitBound>)> {
+    let mut param_constraint_mapping: Vec<(Ident, Vec<TraitBound>)> = type_params
+        .iter()
+        .map(|type_param| {
+            let trait_bounds: Vec<TraitBound> = type_param
+                .bounds
+                .iter()
+                .flat_map(|bound| match bound {
+                    TypeParamBound::Trait(trait_bound) => Some(trait_bound),
+                    _ => None,
+                })
+                .cloned()
+                .collect();
+            (type_param.ident.clone(), trait_bounds)
+        })
+        .collect();
+
+    for predicate in where_clause.predicates.iter() {
+        match predicate {
+            WherePredicate::Type(ref pred_ty) => {
+                let known_bounds: &mut Vec<TraitBound> = match &pred_ty.bounded_ty {
+                    Type::Path(TypePath { path, qself: None }) => {
+                        assert!(
+                            path.leading_colon.is_none(),
+                            "LHS of `where` boundary can't resolve from global scope!"
+                        );
+                        assert_eq!(
+                            1,
+                            path.segments.len(),
+                            "LHS of `where` boundary can't have segments!"
+                        );
```

Comment:
Would this trigger on something like `where crate::Type: Display`? If so we may want to `continue` instead of panic on this case.

Reply:
Yes! I was not very familiar with the kinds of things that could be the subject of `where` constraints on this first attempt--I think it's probable that globally-scoped `::path::to::Type` may also be appropriate to avoid panicking over.

Reply:
Ok, spent a lot of time learning about how `where` works and I think I've made this a lot more clear--thanks so much for this thoughtful remark!